### PR TITLE
MTC-749: margin parameter workflow changed.. it's now by trading inst…

### PR DESCRIPTION
**Pull request by GSailen**
…rument from settings service.

### CODE REVIEW

File: src/MarginTrading.Backend.Contracts/Events/OvernightMarginParameterChangedEvent.cs [hunk 1]
```diff
@@ -22,21 +22,15 @@ public class OvernightMarginParameterChangedEvent
...
-        /// Old value of parameter.
+        /// Current state of parameter.
         /// </summary>
         [Key(2)]
-        public decimal OldValue { get; set; }
+        public bool CurrentState { get; set; }
         
         /// <summary>
-        /// New value of parameter.
+        /// Current value of all multipliers, serialized to JSON.
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Contracts/Events/OvernightMarginParameterChangedEvent.cs]: Why JSON? Use a list of key-value pairs if Dictionary does not work with message pack.

File: src/MarginTrading.Backend.Services/TradingConditions/ITradingInstrumentsCacheService.cs [hunk 2]
```diff
@@ -1,12 +1,19 @@
-﻿using MarginTrading.Backend.Core.TradingConditions;
+﻿using System.Collections.Generic;
+using MarginTrading.Backend.Core.TradingConditions;
 
 namespace MarginTrading.Backend.Services.TradingConditions
 {
     public interface ITradingInstrumentsCacheService
     {
+        void InitCache(IEnumerable<ITradingInstrument> accountAssets);
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/TradingConditions/ITradingInstrumentsCacheService.cs]: Rename accountAssets  -> tradingInstruments please

File: src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs [hunk 3]
```diff
@@ -89,9 +109,88 @@ internal void InitAccountAssetsCache(List<ITradingInstrument> accountAssets)
...
         {
-            return (tradingCondition, instrument);
+            _lockSlim.EnterWriteLock();
+            try
+            {
+                var key = GetInstrumentCacheKey(tradingInstrument.TradingConditionId, tradingInstrument.Instrument);
+
+                _instrumentsCache.Remove(key);
+
+                _instrumentsCache.Add(key, tradingInstrument);
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs]: Remove + Add looks strange )

File: src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs [hunk 4]
```diff
@@ -89,9 +109,88 @@ internal void InitAccountAssetsCache(List<ITradingInstrument> accountAssets)
...
+                _lockSlim.ExitReadLock();
+            }
+        }
+
+        private Dictionary<(string, string), decimal> GetOvernightMarginMultipliers()
+        {
+            _lockSlim.EnterReadLock();
+            try
+            {
+                return _instrumentsCache.ToDictionary(x => x.Key, x => x.Value.OvernightMarginMultiplier);
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs]: I am not sure that we should return all values every time. I would return only the value that changes, that means _instrumentsCache.Where(i => i.Value.OvernightMarginMultiplier != 1).ToDictionary....

File: src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs [hunk 5]
```diff
@@ -89,9 +109,88 @@ internal void InitAccountAssetsCache(List<ITradingInstrument> accountAssets)
...
+
+            if (multiplierChanged)
+            {
+                //send event when the overnight margin parameter is enabled/disabled (margin requirement changes)
+                _cqrsSender.PublishEvent(new OvernightMarginParameterChangedEvent
+                {
+                    CorrelationId = _identityGenerator.GenerateGuid(),
+                    EventTimestamp = _dateService.Now(),
+                    CurrentState = _overnightMarginParameterOn,
+                    MultipliersValue = GetOvernightMarginMultipliers().ToJson(),
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs]: That looks like an error. You should use GetOvernightMarginParameterValues here.

File: src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs (hunk 3 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs]: Why? It's Update method, which adds new item if it didn't exist and changed to the new one if it existed.

File: src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs (hunk 4 above)
Comment (GSailen) [reply src/MarginTrading.Backend.Services/TradingConditions/TradingInstrumentsCacheService.cs]: Lets discuss

**Pull request merged**